### PR TITLE
Pragmas

### PR DESCRIPTION
**Pull request by sebasguts**
I put this on top if the syntax tree PR #114 to see if you can code/decode the pragmas

### CODE REVIEW

File: src/code.c [hunk 1]
```diff
@@ -1806,6 +1806,17 @@ void CodeStringExpr (
     PushExpr( string );
 }
```
Comment (fingolfin) [inline src/code.c:1807]: Commit messsage: "Function can contain pragmas ..." -> Function*s*

File: src/code.c [hunk 2]
```diff
@@ -1806,6 +1806,17 @@ void CodeStringExpr (
...
+void CodePragma (
+    Obj              pragma )
+{
+    GAP_ASSERT(IS_STRING_REP(pragma));
+
+    Expr pragmaexpr = NewStat( T_PRAGMA, sizeof(UInt) );
+    Int ix = PushValue(pragma);
+    WRITE_EXPR(pragmaexpr, 0, ix);
+    PushStat( pragmaexpr );
+}
```
Comment (fingolfin) [inline src/code.c:1817]: This is completely new code, so you are free to clang-format it :-)

File: src/intrprtr.c [hunk 3]
```diff
@@ -2201,6 +2201,19 @@ void           IntrStringExpr (
...
+void           IntrPragma (
+    Obj               pragma )
+{
+    SKIP_IF_RETURNING();
+    SKIP_IF_IGNORING();
+    if ( STATE(IntrCoding)    > 0 ) {
+        CodePragma( pragma );
+    } else {
+        // Push a void when interpreting
+        PushVoidObj();
```
Comment (fingolfin) [inline src/intrprtr.c:2213]: Why is this needed?

File: src/read.c [hunk 4]
```diff
@@ -1474,7 +1474,7 @@ static void ReadLiteral (
     case S_STRING:
         GAP_ASSERT(STATE(ValueObj) != 0);
         TRY_IF_NO_ERROR { IntrStringExpr(STATE(ValueObj)); }
-        Match(S_STRING, "", follow);
+        Match(S_STRING, "", follow );
```
Comment (fingolfin) [inline src/read.c]:
```suggestion
        Match(S_STRING, "", follow);
```

File: src/read.c [hunk 5]
```diff
@@ -2443,7 +2450,12 @@ static UInt ReadStats (
             SyntaxError("statement expected");
         }
         nr++;
-        MatchSemicolon(follow);
+        if( STATE(Symbol) != S_PRAGMA ){
+            MatchSemicolon(follow);
+        }
+        else {
+            Match(S_PRAGMA, "", 0L );
+        }
```
Comment (fingolfin) [inline src/read.c:2458]: What's the reason for this change?

File: src/scanner.c [hunk 6]
```diff
@@ -945,8 +969,17 @@ static UInt NextSymbol(void)
 
     // skip over <spaces>, <tabs>, <newlines> and comments
     while (c == ' ' || c == '\t' || c== '\n' || c== '\r' || c == '\f' || c=='#') {
-        if (c == '#')
+        if (c == '#'){
```
Comment (fingolfin) [inline src/scanner.c]:
```suggestion
        if (c == '#') {
```

File: src/scanner.c [hunk 7]
```diff
@@ -945,8 +969,17 @@ static UInt NextSymbol(void)
...
+            c = GET_NEXT_CHAR_NO_LC();
+            if ( c == '@') {
```
Comment (fingolfin) [inline src/scanner.c]:
```suggestion
            if (c == '@') {
```

File: src/scanner.c [hunk 8]
```diff
@@ -945,8 +969,17 @@ static UInt NextSymbol(void)
...
+                // we have encountered a pragma
+                GetPragma();
+                return S_PRAGMA;
+
```
Comment (fingolfin) [inline src/scanner.c]:
```suggestion
```

File: src/scanner.c [hunk 9]
```diff
@@ -945,8 +969,17 @@ static UInt NextSymbol(void)
 
     // skip over <spaces>, <tabs>, <newlines> and comments
     while (c == ' ' || c == '\t' || c== '\n' || c== '\r' || c == '\f' || c=='#') {
-        if (c == '#')
+        if (c == '#'){
+            c = GET_NEXT_CHAR_NO_LC();
```
Comment (fingolfin) [inline src/scanner.c:973]: I am not sure I understand how this is correct... what if the lines ends right after the `#`? Then this command moves us to the start of the next line, and then `SKIP_TO_END_OF_LINE();` skips *that* line, doesn't it? What am I missing?

File: src/scanner.h [hunk 10]
```diff
@@ -185,7 +187,7 @@ typedef UInt            TypSymbolSet;
                     |S_PLUS|S_MINUS|S_NOT|S_LPAREN)
 
 #define STATBEGIN  (S_IDENT|S_UNBIND|S_IF|S_FOR|S_WHILE|S_REPEAT \
-                    |S_BREAK|S_RETURN|S_HELP|S_QUIT)
+                    |S_BREAK|S_RETURN|S_HELP|S_QUIT|S_PRAGMA)
```
Comment (fingolfin) [inline src/scanner.h]: This is not actually needed, as S_HELP now implies S_PRAGMA

File: src/stats.c [hunk 11]
```diff
@@ -1513,6 +1513,14 @@ static void PrintReturnVoid(Stat stat)
     Pr( "return;", 0L, 0L );
 }
 
+static void PrintPragma(Expr expr)
+{
+    UInt ix = READ_EXPR(expr, 0);
+    Obj string =  GET_VALUE_FROM_CURRENT_BODY(ix);
```
Comment (fingolfin) [inline src/stats.c]:
```suggestion
    Obj string = GET_VALUE_FROM_CURRENT_BODY(ix);
```

File: tst/testinstall/syntaxtree.tst [hunk 12]
```diff
@@ -2854,6 +2861,7 @@ rec(
   stats := rec(
       statements := [ rec(
               obj := rec(
+                  string := "1.0",
                   type := "T_FLOAT_EXPR_EAGER",
                   value := 1 ),
```
Comment (fingolfin) [inline tst/testinstall/syntaxtree.tst:2866]: I am confused as to why `value` is 1 here... I'd expect `1.` or `1.0`... huh

File: src/code.c [hunk 13]
```diff
@@ -1944,7 +1942,7 @@ void CodeFloatExpr(Obj s)
         CodeEagerFloatExpr(s, mark);
     }
     else {
-        CodeLazyFloatExpr(s);
+        CodeLazyFloatExpr(s,1);
```
Comment (fingolfin) [inline src/code.c:1945]:
```suggestion
        CodeLazyFloatExpr(s, 1);
```

File: src/code.c [hunk 14]
```diff
@@ -1909,7 +1904,10 @@ static void CodeLazyFloatExpr(Obj str)
     WRITE_EXPR(fl, 1, PushValue(str));
 
     /* push the expression */
-    PushExpr(fl);
+    if(pushExpr){
```
Comment (fingolfin) [inline src/code.c:1907]:
```suggestion
    if(pushExpr) {
```

File: src/syntaxtree.c [hunk 15]
```diff
@@ -236,15 +333,63 @@ static Obj SyntaxTreeRecExpr(Obj result, Expr expr)
...
     AssPRec(result, RNamName("value"), string);
     return result;
 }
 
+static Obj SyntaxTreeFloatEager(Obj result, Expr expr)
+{
+    Obj value = GET_VALUE_FROM_CURRENT_BODY(READ_EXPR(expr, 0));
+    Obj string = GET_VALUE_FROM_CURRENT_BODY(READ_EXPR(expr, 1));
+    AssPRec(result, RNamName("value"), value);
+    AssPRec(result, RNamName("string"), string);
```
Comment (fingolfin) [inline src/syntaxtree.c:372]: The "mark" character is not decoded here, as returned by `(Char)READ_EXPR(expr, 2)`...

**Review by fingolfin — COMMENTED** [review]
some remarks

File: src/syntaxtree.c [hunk 16]
```diff
@@ -525,12 +781,16 @@ static Int InitLibrary(StructInitInfo * module)
...
     for (UInt tnum = 0; tnum < ARRAY_SIZE(Compilers); tnum++) {
         const char * str = Compilers[tnum].name;
-        if (str)
+        if (str) {
+            AssPRec(typeRec, RNamName(str), ObjInt_UInt(tnum));
             ASS_LIST(typeStrings, tnum + 1, MakeImmString(str));
+        }
     }
 
+
```
Comment (fingolfin) [inline src/syntaxtree.c:793]:
```suggestion
```

File: src/code.c (hunk 1 above)
Comment (sebasguts) [reply src/code.c:1807]: Fixed.

File: src/code.c (hunk 2 above)
Comment (sebasguts) [reply src/code.c:1817]: Done.

File: src/intrprtr.c (hunk 3 above)
Comment (sebasguts) [reply src/intrprtr.c:2213]: If you don't do it, you run into the assertion `assert( LEN_PLIST(STATE(StackObj)) == 1 );` in `IntrEnd` if you enter a pragma in plain interactive mode, i.e., in a GAP session like
```
gap> #@ foo
```

File: src/read.c (hunk 5 above)
Comment (sebasguts) [reply src/read.c:2458]: While a pragma is a statement, you have to `Match` the statement end to get the next symbol. However, a pragma does not end with a semicolon, so we cannot get the next symbol by matching semicolon.

File: src/scanner.c (hunk 9 above)
Comment (sebasguts) [reply src/scanner.c:973]: The problem I am trying to tackling is that there is very specifically [this single comment](https://github.com/gap-system/gap/blob/master/lib/stbcbckt.gi#L207).

But if your current character is a linebreak, `SKIP_TO_END_OF_LINE` does not do anything.

File: src/scanner.h (hunk 10 above)
Comment (sebasguts) [reply src/scanner.h]: Fixed.

File: tst/testinstall/syntaxtree.tst (hunk 12 above)
Comment (sebasguts) [reply tst/testinstall/syntaxtree.tst:2866]: This is part of the other PR #114 and should maybe be discussed there.
